### ISSUE TITLE
arcface on mnist

### ISSUE THREAD

**Issue by johnjwatson** [KevinMusgrave/pytorch-metric-learning#175]
I am trying to run vanilla minst training via `pytorch-metric-learning` but I am not able to get any decent accuracy out of the training/test. So, I tried to use the pytorch packaged (https://github.com/pytorch/examples/tree/master/mnist) mnist training and modified it like so:

```
from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.optim.lr_scheduler import StepLR
from pytorch_metric_learning import losses, regularizers

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout2d(0.25)
        self.dropout2 = nn.Dropout2d(0.5)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        encs_x = self.fc1(x)
        x = F.relu(encs_x)
        x = self.dropout2(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return encs_x, output


def train(args, model, device, train_loader,  loss_func, loss_optimizer, epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        loss_optimizer.zero_grad()
        encs_x, output = model(data)
        loss = loss_func(encs_x, target)
        loss.backward()
        loss_optimizer.step()
        if batch_idx % args.log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.item()))
            if args.dry_run:
                break


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            encs_x, output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))


def main():
    # Training settings
    parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
    parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                        help='input batch size for training (default: 64)')
    parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                        help='input batch size for testing (default: 1000)')
    parser.add_argument('--epochs', type=int, default=50, metavar='N',
                        help='number of epochs to train (default: 14)')
    parser.add_argument('--lr', type=float, default=1.0, metavar='LR',
                        help='learning rate (default: 1.0)')
    parser.add_argument('--gamma', type=float, default=0.7, metavar='M',
                        help='Learning rate step gamma (default: 0.7)')
    parser.add_argument('--no-cuda', action='store_true', default=False,
                        help='disables CUDA training')
    parser.add_argument('--dry-run', action='store_true', default=False,
                        help='quickly check a single pass')
    parser.add_argument('--seed', type=int, default=1, metavar='S',
                        help='random seed (default: 1)')
    parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                        help='how many batches to wait before logging training status')
    parser.add_argument('--save-model', action='store_true', default=True,
                        help='For Saving the current Model')
    args = parser.parse_args()
    # use_cuda = not args.no_cuda and torch.cuda.is_available()

    torch.manual_seed(args.seed)

    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    print("==> COMPUTATION DEVICE IS:", device)

    # kwargs = {'batch_size': args.batch_size}
    # # if use_cuda:
    # kwargs.update({'num_workers': 1,
    #                 'pin_memory': True,
    #                 'shuffle': True},
    #                 )

    transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
    dataset1 = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
    dataset2 = datasets.MNIST('../data', train=False,
                       transform=transform)
    train_loader = torch.utils.data.DataLoader(dataset1,batch_size=64, shuffle=True)   #**kwargs)
    test_loader = torch.utils.data.DataLoader(dataset2,batch_size=1, shuffle=False) #, **kwargs)

    model = Net().to(device)

    R = regularizers.RegularFaceRegularizer()
    loss_func = losses.ArcFaceLoss(10, 128, margin=28.6, scale=64, weight_regularizer=R).to(torch.device('cuda'))
    params = list(model.parameters()) + list(loss_func.parameters())
    loss_optimizer = torch.optim.Adam(params, lr=0.01)

    scheduler = StepLR(loss_optimizer, step_size=1, gamma=args.gamma)
    for epoch in range(1, args.epochs + 1):
        train(args, model, device, train_loader, loss_func, loss_optimizer, epoch)
        test(model, device, test_loader)
        scheduler.step()

    if args.save_model:
        torch.save(model.state_dict(), "mnist_cnn.tar")


if __name__ == '__main__':
    main()
```

I am not excatly sure if what I am doing is correct, but the goal is to get the embeddings for the classifier. Any clues as to what I am doing wrong would be great. I get something around 18% accuracy, which does not seem right :((

PS: great package!

**Comment by KevinMusgrave** [comment]
It looks like during training, you aren't using ```output```, i.e. you compute the loss using ```encs_x```. Therefore, the layers after ```encs_x``` are unaffected by backpropagation. But then you use ```output``` during test time, even though it has not been trained at all. 

One way to fix this is to just remove fc2:
```python
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout2d(0.25)
        self.dropout2 = nn.Dropout2d(0.5)
        self.fc1 = nn.Linear(9216, 128)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x= self.fc1(x)
        x = F.relu(x)
        return self.dropout2(x)
```

Then in your training loop:
```
        output = model(data)
        loss = loss_func(output, target)
```

Then during testing you have 2 choices. One is to use nearest neighbors classification based on the embedding space (which is more along the lines of "metric learning"), or you can use argmax like you've already done. But it's a bit more complicated because that part of the model is actually in the loss function now, and I haven't yet written an easy way to get the ArcFace classification result. Anyway, this might work (I haven't tested it):
```python
            output = model(data)
            #### the next few lines are just copied from ArcFaceLoss ####
            mask = loss_func.get_target_mask(output, labels)
            cosine = loss_func.get_cosine(output)
            cosine_of_target_classes = cosine[mask == 1]
            modified_cosine_of_target_classes = loss_func.modify_cosine_of_target_classes(cosine_of_target_classes, cosine, output, labels, mask)
            diff = (modified_cosine_of_target_classes - cosine_of_target_classes).unsqueeze(1)
            logits = cosine + (mask*diff)
            logits = self.scale_logits(logits, output)
            pred = logits.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
```

**Comment by johnjwatson** [comment]
@KevinMusgrave this is awesome - the logits approach worked out of the box! On a related note, I would like to try the metric-learning approach too (so that I can get use dto your awesome package!), but the examples you have shown appear a bit convoluted(pun intended). Is there a way I can simply pass the embedded vectors to the `GlobalEmbeddingSpaceTester` along with the labels to get say the indices of the nearest neighbours? Sure I can install faiss or annoy to do this, but given you allready have classes for this, it would be great if I can just plug the vectors and labels for the classification result. Perhaps there is a snippet somewhere to do this?

**Comment by KevinMusgrave** [comment]
Something like this should work:
```python
from pytorch_metric_learning.testers import GlobalEmbeddingSpaceTester

tester = GlobalEmbeddingSpaceTester()

# dictionary mapping from split names to the datasets
# you can name them whatever you want
# you can also just pass in a single dataset
dataset_dict = {"train": training_set, "val": val_set} 

epoch = #set this to something, it's just for logging purposes

tester.test(dataset_dict, epoch, model)

# This will print various accuracy metrics for both all the splits in dataset_dict
print(tester.all_accuracies)
```

That's probably the easiest way to get accuracy. You can make it fancier by add logging and an embedding space visualizer if you want.

The inference module is supposed to be for deployment, i.e. when you want to compare just two images and determine if they're the same class, or retrieve the nearest images given a single query image.

**Comment by johnjwatson** [comment]
@KevinMusgrave thank you so much for this snippet. Works as expected. May I ask: when you invoke:
`tester.test(dataset_dict, epoch, model)`, I think you do the following:
[1] get the embeddings from the model
[2] get knn for the train and test sets (if both are specified) from the embeddings and the labels and compute accuracy/recall.

Is this roughly what you do here? I ask because I think I can use these in other projects without any friction :+1:

**Comment by KevinMusgrave** [comment]
Yep that's it. 

Note that it doesn't compute classification accuracy (i.e. the predicted class based on a majority vote of neighbors), but if you want that you can easily extend the accuracy calculator. See the docs: https://kevinmusgrave.github.io/pytorch-metric-learning/accuracy_calculation/

There's also various options for the tester: https://kevinmusgrave.github.io/pytorch-metric-learning/testers/

**Comment by johnjwatson** [comment]
@KevinMusgrave thank you again, I will give that a spin today :)
Thanks again for the awesome package! incredibly useful!

**Comment by KevinMusgrave** [comment]
@johnjwatson I was just thinking about implementing a "return logits" function, and realized that the solution I provided above doesn't make sense for inference because it assumes the label is known for each input. Sorry about that!
